packages/playwright-crawler/src/internals/adaptive-playwright-crawler.ts

```diff
@@ -0,0 +1,263 @@
+            const plainHTTPRun = await this.runRequestHandlerWithPlainHTTP(crawlingContext);
+
+            const detectionResult: RenderingType = (() => {
+                if (!plainHTTPRun.ok) {
+                    return 'clientOnly';
+                }
+
+                if (this.resultComparator(plainHTTPRun.result, browserRun.result)) {
+                    return 'static';
+                }
+
+                return 'clientOnly';
+            })();
+
+            crawlingContext.log.info(`Detected rendering type ${detectionResult} for ${crawlingContext.request.url}`);
+            this.renderingTypePredictor.storeResult(url, detectionResult);
+        }
+    }
+
+    protected async commitResult(
+        crawlingContext: PlaywrightCrawlingContext<Dictionary>,
+        { datasetItems, enqueuedUrls, requestUrls, requestListUrls, keyValueStoreChanges }: RequestHandlerResult,
+    ): Promise<void> {
+        for (const {item, datasetIdOrName} of datasetItems) {
+            await crawlingContext.pushData(item, datasetIdOrName)
+        }
+        for (const {options} of enqueuedUrls) {
+            await crawlingContext.enqueueLinks(options)
+        }
+        for (const {request, options} of requestUrls) {
+            await crawlingContext.addRequests([request], options)
+        }
+        for (const {request, options} of requestListUrls) {
+            await crawlingContext.addRequests([request], options)
+        }
+        for (const [storeIdOrName, changes] of Object.entries(keyValueStoreChanges)) {
+            const store = await crawlingContext.getKeyValueStore(storeIdOrName);
+            for (const [key, { changedValue, options }] of Object.entries(changes)) {
+                await store.setValue(key, changedValue, options)
+            }
+        }
+    }
+
+    protected async runRequestHandlerInBrowser(crawlingContext: PlaywrightCrawlingContext<Dictionary>): Promise<Result<RequestHandlerResult>> {
+        const result = new RequestHandlerResult(this.config);
+
+        try {
+            await super._runRequestHandler.call(
+                new Proxy(this, {
+                    get: (target, propertyName, receiver) => {
+                        if (propertyName === 'userProvidedRequestHandler') {
+                            return (playwrightContext: PlaywrightCrawlingContext) => this.adaptiveRequestHandler({
+                                request: crawlingContext.request,
+                                log: crawlingContext.log,
+                                dom: playwrightLocatorPortadom(playwrightContext.page.locator(':root'), playwrightContext.page),
+                                enqueueLinks: async (options = {}) => {
+                                    const $ = await playwrightContext.parseWithCheerio();
+                                    const urls = extractUrlsFromCheerio($, options.selector, playwrightContext.request.loadedUrl); // TODO avoid parsing with cheerio
```

Comment:
I'll probably have to pull `extractUrlsFromPage` from `browser-crawler.ts` to some utils file, like I did with the Cheerio version. Is that OK?